### DISCUSSION TITLE
Plans for antibody-derived tags (CITE-seq) and other data modalities

### DISCUSSION THREAD

**Discussion by PeteHaitch** [kanaverse/kana#119] [Ideas]
We're finding kana really useful in our collaborative work and my colleagues are starting to use it instead of or to supplement interactive analyses they usually perform with 10x Genomics' Loupe Browser.
One feature that I think **kana** currently lacks, in comparison to the Loupe Browser, is the ability to analyse (or even visualise?) data from other modalities, such as antibody-derived tags (ADTs) used in CITE-seq.
Is this something that you are considering?

An issue I foresee is how these data are stored:

- 10x: these features are simply additional rows of the count matrix that are annotated as `Antibody Capture` rather than `Gene Expression`
- H5AD: I don't think H5AD files have the notion of an 'alternative experiment', which is where such data would typically live in a _SingleCellExperiment_-based workflow. I might be wrong, but it seems like they (the Python/AnnData/scanpy single-cell world) are using [**MuData**](https://mudata.readthedocs.io/en/latest/), so that would be yet another format to support (although it at least seems to be similar to`.h5ad`).

Keen to hear your thoughts on this.

Cheers,
Pete
CC: @ashsea1

**Comment by LTLA** [comment]
Uh... well, no plans, really.

Right now, if you're using the 10X formats, they'll just get treated as extra genes, which may or may not be sensible. They should at least be visualize-able somewhere, you can just search for them in the marker table.

Actually making the analysis aware of ADTs would be a bit of effort. Not impossible, but it would require some thought. The cheapest approach would be to weight each feature somehow (e.g., using **mumosa**'s neighbor-based weighting scheme) and let the PCA handle the rest. Weights could be adjusted to be gene-only, ADT-only, equivalent weights, or something in between, if users want to toy around with the effects of changing the contributions of the two modalities on the clustering/t-SNE/UMAP output.

**Reply by PeteHaitch** [reply]
Yeah I appreciate that doing it 'properly' would be a fair effort. 
Would you consider an 'analyse GEX but with capability to overlay ADT data' approach (e.g., overlay expression of selected ADT on GEX-derived t-SNE/UMAP)?

**Reply by LTLA** [reply]
Yes, short-term, that could probably be done relatively quickly by shoveling those features into the column annotations and subsetting the matrix by feature in **bakana**'s 10X reader(s). Downstream steps would then be gene-only and the ADT expression would show up in the legend without any further UI changes. I would still need to add some ADT-aware normalization to **scran.js**, though.

**Reply by PeteHaitch** [reply]
Happy to help out if I can

**Reply by LTLA** [reply]
It's been a while since I looked at ADT data, so I don't know what the modern panel designs are.

Are Ig controls standard these days? If so, we could just use those to normalize, e.g., take the sum of the control counts and use those as the size factors (any cells with zero counts are just dropped as part of QC).

If not, we'll need to do something wild to deal with the composition biases. I don't like the geometric approach much. So instead:

- Cluster on the ADT counts, e.g., using 20-30 k-means clusters.
- Do median-based normalization across the cluster centroids, exploiting their relative density to get well-defined size factors.
- Do library size normalization within each cluster, and take the product to obtain the size factor for each cell.

**Reply by PeteHaitch** [reply]
We've been regularly using BioLegend's TotalSeq Human and Mouse Universal Cocktail panels, which do include isotype control antibodies:

- [TotalSeq™-A Human Universal Cocktail, V1.0](https://www.biolegend.com/nl-be/products/totalseq-a-human-universal-cocktail-v1-0-20321) includes 9 isotype control antibodies
- [TotalSeq™-C Human Universal Cocktail, V1.0](https://www.biolegend.com/en-gb/products/totalseq-c-human-universal-cocktail-v1-0-19736) includes 7 isotype control antibodies
- [TotalSeq™-A Mouse Universal Cocktail, V1.0](https://www.biolegend.com/en-us/products/totalseq-a-mouse-universal-cocktail-v1-0-21733) includes 9 isotype control antibodies

However, people also use custom panels that don't necessarily include such controls.

**Reply by LTLA** [reply]
The normalization methods discussed above have been added to **libscran**, see the latest few commits on [`master`](https://github.com/LTLA/libscran/commits/master). From there, they'll begin their slow journey through the stack, i.e., **scran.js**, then **bakana**, then **kana** itself.

**Reply by jkanche** [reply]
its now up on kana

**Reply by PeteHaitch** [reply]
Awesome, we'll take it for a spin!

**Reply by PeteHaitch** [reply]
From what I can tell, support has been added for the 10x and mtx readers.
Would it be feasible to add support to the h5ad reader?

**Reply by LTLA** [reply]
Hard to say. I don't believe there's a consistent field for the feature type in their row annotations, so we would have to guess. We could hope that whoever generated the H5AD would have named it `feature_type`, but that's not very reliable. I suppose it's no less reliable than our guessing of the gene symbols, but even that just involves checking if there's a `name` or `sym(bol)` column somewhere.

**Reply by PeteHaitch** [reply]
Yeah ... I realised that after posting. I'll ponder it over the weekend.

**Reply by jkanche** [reply]
@PeteHaitch our current release (v2.0.0) might be of interest to you. we now fully support multi-modal analysis of ADT + RNA

**Reply by PeteHaitch** [reply]
Awesome, thank you both! And just in time for a talk I'm giving today where I'm promoting **kana**.
